src/server/ua_file.c

```diff
@@ -0,0 +1,419 @@
+
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldOpenQN, &openNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldReadQN, &readNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldSetPositionQN, &setPositionNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldGetPositionQN, &getPositionNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldWriteQN, &writeNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldCloseQN, &closeNodeId);
+
+    result |= UA_Server_setMethodNode_callback(server, openNodeId, UA_Server_openFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, readNodeId, UA_Server_readFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, setPositionNodeId, UA_Server_setPositionFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, getPositionNodeId, UA_Server_getPositionFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, writeNodeId, UA_Server_writeFileCallback);
+    result |= UA_Server_setMethodNode_callback(server, closeNodeId, UA_Server_closeFileCallback);
+
+    return result;
+}
+
+UA_StatusCode
+setFileTypeInfo(UA_Server *server, const UA_NodeId fileTypeObjectNodeId,
+                          UA_String filePath)
+{
+    UA_FileType *fileTypeObject = (UA_FileType*) UA_malloc(sizeof(UA_FileType));
+    UA_NodeId_copy(&fileTypeObjectNodeId, &fileTypeObject->fileTypeObjectNodeId);
+    UA_String_copy(&filePath, &fileTypeObject->filePath);
+    fileTypeObject->openCount = 0;
+    fileTypeObject->fileSize = 0;
+    fileTypeObject->writable = UA_TRUE;
+    fileTypeObject->userWritable = UA_TRUE;
+    setFileObjectNodeProperties(server, fileTypeObject);
+    LIST_INIT(&fileTypeObject->fileInfo);
+    LIST_INSERT_HEAD(&server->fileObjects, fileTypeObject, listEntry);
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode 
+UA_Server_addFile(UA_Server *server, const UA_NodeId requestedNewNodeId,
+                  const UA_NodeId parentNodeId,
+                  const UA_NodeId referenceTypeId,
+                  const UA_QualifiedName browseName,
+                  const UA_ObjectAttributes attr,
+                  const UA_String filePath,
+                  void *nodeContext, UA_NodeId *outNewNodeId)
+{
+    UA_StatusCode result = UA_Server_addObjectNode(server, requestedNewNodeId,
+                            parentNodeId, referenceTypeId, browseName,
+                            UA_NODEID_NUMERIC(0, UA_NS0ID_FILETYPE),
+                            attr, nodeContext, outNewNodeId);
+    
+    UA_file *sampleFile = UA_file_open((char*)filePath.data, "w");
+    UA_file_close(sampleFile);
+
+    //If NodeId is assigned by server
+    if(outNewNodeId)
+    {
+        result |= setFileTypeInfo(server, *outNewNodeId, filePath);
+        result |= setFileMethodCallbacks(server, *outNewNodeId);
+    }
+
+    //else use the  NodeId requested by user
+    else
```

Comment:
This will fail if the user does not provider outNewNodeId and if the NodeId has `identifier.numeric==0`.
That is the convention to ask the nodestore to generate a new numeric nodeid.
Use a temporary newNodeId variable.